File: crates/defi-core/src/traits/gauge.rs
use alloy::primitives::{Address, U256};
use async_trait::async_trait;

use crate::error::Result;
use crate::types::{DeFiTx, RewardInfo};

/// ve(3,3) Gauge operations — stake LP tokens to earn emissions
#[async_trait]
pub trait Gauge: Send + Sync {
    fn name(&self) -> &str;

    /// Deposit LP tokens into gauge
    async fn build_deposit(
        &self,
        gauge: Address,
        amount: U256,
        token_id: Option<U256>,
    ) -> Result<DeFiTx>;

    /// Withdraw LP tokens from gauge
    async fn build_withdraw(&self, gauge: Address, amount: U256) -> Result<DeFiTx>;

    /// Claim earned rewards from gauge
    async fn build_claim_rewards(&self, gauge: Address) -> Result<DeFiTx>;

    /// Get pending rewards for a user
    async fn get_pending_rewards(&self, gauge: Address, user: Address) -> Result<Vec<RewardInfo>>;
}

/// ve(3,3) Vote-escrow operations — lock tokens for veNFT
#[async_trait]
pub trait VoteEscrow: Send + Sync {
    fn name(&self) -> &str;

    /// Create a new veNFT lock
    async fn build_create_lock(&self, amount: U256, lock_duration: u64) -> Result<DeFiTx>;

    /// Increase lock amount
    async fn build_increase_amount(&self, token_id: U256, amount: U256) -> Result<DeFiTx>;

    /// Increase lock duration
    async fn build_increase_unlock_time(
        &self,
        token_id: U256,
        lock_duration: u64,
    ) -> Result<DeFiTx>;

    /// Withdraw after lock expires
    async fn build_withdraw_expired(&self, token_id: U256) -> Result<DeFiTx>;
}

/// ve(3,3) Voter operations — vote on gauge emissions
#[async_trait]
pub trait Voter: Send + Sync {
    fn name(&self) -> &str;

    /// Vote for gauges with veNFT
    async fn build_vote(
        &self,
        token_id: U256,
        pools: Vec<Address>,
        weights: Vec<U256>,
    ) -> Result<DeFiTx>;

    /// Claim bribes for voted pools
    async fn build_claim_bribes(&self, bribes: Vec<Address>, token_id: U256) -> Result<DeFiTx>;

    /// Claim trading fees
    async fn build_claim_fees(&self, fees: Vec<Address>, token_id: U256) -> Result<DeFiTx>;
}

/// Combined ve(3,3) system — gauge staking + vote-escrow + voter
///
/// Implementors of this trait provide the full ve(3,3) stack.
/// The trait is auto-implemented for any type implementing all three sub-traits.
pub trait GaugeSystem: Gauge + VoteEscrow + Voter {}

impl<T: Gauge + VoteEscrow + Voter> GaugeSystem for T {}

File: ts/packages/defi-core/src/traits/gauge.ts
import type { Address } from "viem";
import type { DeFiTx, GaugedPool, RewardInfo } from "../types.js";

/** ve(3,3) Gauge operations — stake LP tokens to earn emissions */
export interface IGauge {
  name(): string;
  /** Resolve gauge address from pool address via voter */
  resolveGauge?(pool: Address): Promise<Address>;
  /** Deposit LP tokens into gauge */
  buildDeposit(gauge: Address, amount: bigint, tokenId?: bigint): Promise<DeFiTx>;
  /** Withdraw LP tokens or NFT from gauge */
  buildWithdraw(gauge: Address, amount: bigint, tokenId?: bigint): Promise<DeFiTx>;
  /** Claim earned rewards from gauge */
  buildClaimRewards(gauge: Address, account?: Address): Promise<DeFiTx>;
  /** Claim rewards for a CL gauge NFT position (Hybra V4 style) */
  buildClaimRewardsByTokenId?(gauge: Address, tokenId: bigint): Promise<DeFiTx>;
  /** Get pending rewards for a user */
  getPendingRewards(gauge: Address, user: Address): Promise<RewardInfo[]>;
  /** Get pending rewards for a CL gauge NFT position */
  getPendingRewardsByTokenId?(gauge: Address, tokenId: bigint): Promise<bigint>;
  /** Discover all pools that have active emission gauges */
  discoverGaugedPools?(): Promise<GaugedPool[]>;
}

/** ve(3,3) Vote-escrow operations — lock tokens for veNFT */
export interface IVoteEscrow {
  name(): string;
  /** Create a new veNFT lock */
  buildCreateLock(amount: bigint, lockDuration: number): Promise<DeFiTx>;
  /** Increase lock amount */
  buildIncreaseAmount(tokenId: bigint, amount: bigint): Promise<DeFiTx>;
  /** Increase lock duration */
  buildIncreaseUnlockTime(tokenId: bigint, lockDuration: number): Promise<DeFiTx>;
  /** Withdraw after lock expires */
  buildWithdrawExpired(tokenId: bigint): Promise<DeFiTx>;
}

/** ve(3,3) Voter operations — vote on gauge emissions */
export interface IVoter {
  name(): string;
  /** Vote for gauges with veNFT */
  buildVote(tokenId: bigint, pools: Address[], weights: bigint[]): Promise<DeFiTx>;
  /** Claim bribes for voted pools */
  buildClaimBribes(bribes: Address[], tokenId: bigint): Promise<DeFiTx>;
  /** Claim trading fees */
  buildClaimFees(fees: Address[], tokenId: bigint): Promise<DeFiTx>;
}

/** Combined ve(3,3) system — gauge staking + vote-escrow + voter */
export interface IGaugeSystem extends IGauge, IVoteEscrow, IVoter {}
